src/app/technical-record/technical-record.component.ts

```diff
@@ -1,116 +1,116 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class TechnicalRecordComponent implements OnInit {
+  showAdrDetails: boolean;
+  adrDisplayParams: { [key: string]: boolean };
+  techRecord: FormGroup;
+
+  allOpened = false;
+  panels: { panel: string; isOpened: boolean }[] = [
+    { panel: 'panel1', isOpened: false },
+    { panel: 'panel2', isOpened: false },
+    { panel: 'panel3', isOpened: false },
+    { panel: 'panel4', isOpened: false },
+    { panel: 'panel5', isOpened: false },
+    { panel: 'panel6', isOpened: false },
+    { panel: 'panel7', isOpened: false },
+    { panel: 'panel8', isOpened: false },
+    { panel: 'panel9', isOpened: false },
+    { panel: 'panel10', isOpened: false }
+  ];
+
+  @Input() vehicleTechRecord: VehicleTechRecordModel;
+  @Input() activeRecord: TechRecord;
+  @Input() metaData: MetaData;
+  @Input() editState: VIEW_STATE;
+  @Input() testResultJson: TestResultModel[];
+  @Output() submitTechRecord = new EventEmitter<TechRecord>();
+  @Output() changeViewState = new EventEmitter<VIEW_STATE>();
+
+  constructor(
+    private dialog: MatDialog,
+    private allowedValues: TechnicalRecordValuesMapper,
+    public techRecHelpers: TechRecordHelpersService
+  ) {}
+
+  ngOnInit() {
+    this.techRecord = new FormGroup({});
+    this.adrDisplayParams = { showAdrDetails: !!this.activeRecord.adrDetails };
+  }
+
+  togglePanel() {
+    for (const panel of this.panels) {
+      panel.isOpened = !this.allOpened;
+    }
+    this.allOpened = !this.allOpened;
+  }
+
+  hasSecondaryVrms(vrms) {
+    return vrms && vrms.length > 1 && vrms.filter((vrm) => vrm.isPrimary === false).length > 0;
+  }
+
+  editTechRecord() {
+    this.changeViewState.emit(VIEW_STATE.EDIT);
+  }
+
+  cancelTechRecordEdit() {
+    this.changeViewState.emit(VIEW_STATE.VIEW_ONLY);
+    this.activeRecord = JSON.parse(JSON.stringify(this.activeRecord));
+    this.adrDisplayParams = { ...{ showAdrDetails: !!this.activeRecord.adrDetails } };
+
+    const controlNames = Object.keys(this.techRecord.controls);
+    for (let index = 0; index < controlNames.length; index++) {
+      this.techRecord.removeControl(controlNames[index]);
+    }
+    this.techRecord.updateValueAndValidity();
+  }
+
+  onSaveChanges({ valid, value }: { valid: boolean; value: any }) {
+    // if (valid) { TODO: Re-enable during ADR Validation ticket
+
+    const editedRecord: TechRecord = this.allowedValues.mapControlValuesToDataValues(
+      JSON.parse(JSON.stringify(value))
+    );
+    const mergedRecord: TechRecord = { ...this.activeRecord, ...editedRecord };
+
+    const dialogRef = this.dialog.open(AdrReasonModalComponent, {
+      width: '600px',
+      data: {
+        context: 'Enter reason for changing technical record'
+      }
+    });
+
+    dialogRef.afterClosed().subscribe((result) => {
+      if (result && result.isSave) {
+        mergedRecord.reasonForCreation = result.data;
+        this.submitTechRecord.emit(mergedRecord);
```

Comment:
I think it would make things easier if we use ngrx and showed in in the state instead? What do you think?

Reply:
Yes I agree this can be solved in NGRX way. We would need to extract this implementation and the dialog component implementation and solve in NGRX way. Not sure on when or how. I will discuss with you @JonLazariniDD